### PR TITLE
fix: gpkg zoom and extent update

### PR DESCRIPTION
**Pull request by cl-shaharshavit-rnd**

### CODE REVIEW

File: MergerLogic/DataTypes/Gpkg.cs [hunk 1]
```diff
@@ -42,47 +27,34 @@ public class Gpkg : Data<IGpkgUtils>
...
-        {
-            Extent baseExtent = this.utils.GetExtent();
-            Extent extent = gpkg.utils.GetExtent();
-            Extent combinedExtent = new Extent();
-
-            combinedExtent.minX = Math.Min(baseExtent.minX, extent.minX);
-            combinedExtent.minY = Math.Min(baseExtent.minY, extent.minY);
-            combinedExtent.maxX = Math.Max(baseExtent.maxX, extent.maxX);
-            combinedExtent.maxY = Math.Max(baseExtent.maxY, extent.maxY);
+                if (extent is null)
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Gpkg.cs:37]: Why? you just checked this before.

File: MergerLogic/DataTypes/Gpkg.cs [hunk 2]
```diff
@@ -42,47 +27,34 @@ public class Gpkg : Data<IGpkgUtils>
...
+                }
 
-            this.utils.UpdateExtent(combinedExtent);
+                if (!this.utils.Exist())
+                {
+                    this.utils.Create(extent.Value, isOneXOne);
+                }
+                else
+                {
+                    this.utils.DeleteTileTableTriggers();
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Gpkg.cs:49]: Why in else? Create also adds all triggers.

File: MergerLogic/Utils/GpkgUtils.cs [hunk 3]
```diff
@@ -721,5 +664,27 @@ public bool Exist()
...
+
+        public void UpdateTileMatrixTable(bool isOneXOne = false)
+        {
+            using (var connection = new SQLiteConnection($"Data Source={this.path}"))
+            {
+                int maxZoom;
+                connection.Open();
+                using (var command = connection.CreateCommand())
+                {
+                    command.CommandText = $"SELECT  MAX(zoom_level) AS maxZoom FROM \"{this._tileCache}\"; ";
```
Comment (shimoncohen) [inline MergerLogic/Utils/GpkgUtils.cs:676]: Has extra space character in SQL.

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: MergerLogic/DataTypes/Gpkg.cs (hunk 2 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Gpkg.cs:49]: create dont add those triggers

File: MergerLogic/Utils/GpkgUtils.cs [hunk 4]
```diff
@@ -556,14 +501,15 @@ private void SetPragma(SQLiteConnection connection)
...
+        private void CreateSqureGrid(SQLiteConnection connection, int minZoom, int maxZoom, int baseWidth, int baseHeight, int yAxisSizeDeg, int zoomMultipiler, int tileSize)
         {
-            StringBuilder gridBuilder = new StringBuilder("INSERT INTO \"gpkg_tile_matrix\" VALUES ");
+            StringBuilder gridBuilder = new StringBuilder("INSERT OR REPLACE INTO \"gpkg_tile_matrix\" VALUES ");
 
-            int width = baseWidth;
-            int height = baseHeight;
-            double res = baseRes;
-            for (int z = 0; z <= maxZoom; z++)
+            int startZoomMultiplier = (int)Math.Pow(zoomMultipiler, minZoom);
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/GpkgUtils.cs:508]: the res calc is 1 line.
this is the size calc

File: MergerLogic/Utils/GpkgUtils.cs [hunk 5]
```diff
@@ -580,28 +526,24 @@ private void CreateSqureGrid(SQLiteConnection connection, int maxZoom, int baseW
...
             {
                 command.CommandText = "INSERT INTO \"gpkg_tile_matrix_set\" VALUES " +
                     $"('{this._tileCache}',{GeoUtils.SRID},-180,-90,180,90);";
                 command.ExecuteNonQuery();
             }
-            this.CreateSqureGrid(connection, maxZoom, 2, 1, TWO_X_ONE_ZOOM_0_RES, 2, 256);//creates 2X1 grid
         }
 
-        private void Add1X1Data(SQLiteConnection connection, int maxZoom)
+        private void Add1X1MatrixSet(SQLiteConnection connection)
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/GpkgUtils.cs:539]: I think this is cleaner and more readable then defining the extent high up the stack

File: MergerLogic/Utils/GpkgUtils.cs [hunk 6]
```diff
@@ -650,6 +592,7 @@ public void CreateTileCacheValidationTriggers()
         {
             using (var connection = new SQLiteConnection($"Data Source={this.path}"))
             {
+                connection.Open();
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/GpkgUtils.cs:595]: yes

File: MergerLogic/DataTypes/Gpkg.cs (hunk 1 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Gpkg.cs:37]: this meant to be refactor of the previous line.
the first error was removed

File: MergerLogic/Utils/GpkgUtils.cs (hunk 3 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/GpkgUtils.cs:676]: fixed

File: MergerLogic/DataTypes/Gpkg.cs [hunk 7]
```diff
@@ -42,47 +27,34 @@ public class Gpkg : Data<IGpkgUtils>
             {
                 this._coordsFromCurrentGrid = cords => cords;
             }
-            if (!this.utils.Exist() && isBase && maxZoom is not null && extent is not null)
+            if (isBase && extent is null)
             {
-                this.utils.Create(extent.Value, maxZoom.Value, isOneXOne);
+                //throw error if extent is missing in base
+                throw new Exception($"path '{path}' to data does not exist.");
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Gpkg.cs:33]: fixed

File: MergerLogic/DataTypes/Gpkg.cs [hunk 8]
```diff
@@ -42,47 +27,34 @@ public class Gpkg : Data<IGpkgUtils>
...
-            Extent combinedExtent = new Extent();
-
-            combinedExtent.minX = Math.Min(baseExtent.minX, extent.minX);
-            combinedExtent.minY = Math.Min(baseExtent.minY, extent.minY);
-            combinedExtent.maxX = Math.Max(baseExtent.maxX, extent.maxX);
-            combinedExtent.maxY = Math.Max(baseExtent.maxY, extent.maxY);
+                if (extent is null)
+                {
+                    //throw error if extent is missing in base
+                    throw new Exception($"path '{path}' to data does not exist.");
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Gpkg.cs:40]: fixed

File: MergerLogic/DataTypes/Gpkg.cs (hunk 2 above)
Comment (shimoncohen) [reply MergerLogic/DataTypes/Gpkg.cs:44]: I looked in the code, it seemed like it does. Check again to be sure.

Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Gpkg.cs:44]: it doesn't there are 2 tables that has triggers. it isn't the same triggers

**Review by asafMasa — APPROVED** [review]

**Review by shimoncohen — APPROVED** [review]

**Pull request merged**